File: java/src/kanzi/entropy/RangeEncoder.java
/*
Copyright 2011-2017 Frederic Langlet
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
you may obtain a copy of the License at

                http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package kanzi.entropy;

import kanzi.EntropyEncoder;
import kanzi.OutputBitStream;


// Based on Order 0 range coder by Dmitry Subbotin itself derived from the algorithm
// described by G.N.N Martin in his seminal article in 1979.
// [G.N.N. Martin on the Data Recording Conference, Southampton, 1979]
// Optimized for speed.

// Not thread safe
public final class RangeEncoder implements EntropyEncoder
{
    private static final long TOP_RANGE    = 0x0FFFFFFFFFFFFFFFL;
    private static final long BOTTOM_RANGE = 0x0000000000FFFFFFL;
    private static final long RANGE_MASK   = 0x0FFFFF0000000000L;
    private static final int DEFAULT_CHUNK_SIZE = 1 << 16; // 64 KB by default
    private static final int DEFAULT_LOG_RANGE = 13;


    private long low;
    private long range;
    private final int[] alphabet;
    private final int[] freqs;
    private final long[] cumFreqs;
    private final EntropyUtils eu;
    private final OutputBitStream bitstream;
    private final int chunkSize;
    private final int logRange;
    private int shift;
    
    
    public RangeEncoder(OutputBitStream bitstream)
    {
       this(bitstream, DEFAULT_CHUNK_SIZE, DEFAULT_LOG_RANGE);
    }
    
    
    // The chunk size indicates how many bytes are encoded (per block) before 
    // resetting the frequency stats. 0 means that frequencies calculated at the
    // beginning of the block apply to the whole block.
    // The default chunk size is 65536 bytes.
    public RangeEncoder(OutputBitStream bs, int chunkSize, int logRange)
    {
      if (bs == null)
         throw new NullPointerException("Invalid null bitstream parameter");

      if ((chunkSize != 0) && (chunkSize < 1024))
         throw new IllegalArgumentException("The chunk size must be at least 1024");

      if (chunkSize > 1<<30)
         throw new IllegalArgumentException("The chunk size must be at most 2^30");

      if ((logRange < 8) || (logRange > 16))
         throw new IllegalArgumentException("Invalid range parameter: "+logRange+" (must be in [8..16])");

      this.bitstream = bs;
      this.alphabet = new int[256];
      this.freqs = new int[256];
      this.cumFreqs = new long[257];
      this.logRange = logRange;
      this.chunkSize = chunkSize;
      this.eu = new EntropyUtils();
    }

    
   protected int updateFrequencies(int[] frequencies, int size, int lr)
   {
      if ((frequencies == null) || (frequencies.length != 256))
         return -1;

      int alphabetSize = this.eu.normalizeFrequencies(frequencies, this.alphabet, size, 1<<lr);
      
      if (alphabetSize > 0)
      {
         this.cumFreqs[0] = 0;

         // Create histogram of frequencies scaled to 'range'
         for (int i=0; i<256; i++)
            this.cumFreqs[i+1] = this.cumFreqs[i] + frequencies[i];
      }
          
      this.encodeHeader(alphabetSize, this.alphabet, frequencies, lr);
      return alphabetSize;
   }

   
   protected boolean encodeHeader(int alphabetSize, int[] alphabet, int[] frequencies, int lr)
   {
      EntropyUtils.encodeAlphabet(this.bitstream, alphabet, alphabetSize);

      if (alphabetSize == 0)
         return true;

      this.bitstream.writeBits(lr-8, 3); // logRange
      int inc = (alphabetSize > 64) ? 16 : 8;
      int llr = 3;

      while (1<<llr <= lr)
         llr++;

      // Encode all frequencies (but the first one) by chunks of size 'inc'
      for (int i=1; i<alphabetSize; i+=inc)
      {
         int max = 0;
         int logMax = 1;
         final int endj = (i+inc < alphabetSize) ? i + inc : alphabetSize;

         // Search for max frequency log size in next chunk
         for (int j=i; j<endj; j++)
         {
            if (frequencies[alphabet[j]] > max)
               max = frequencies[alphabet[j]];
         }

         while (1<<logMax <= max)
            logMax++;

         this.bitstream.writeBits(logMax-1, llr);

         // Write frequencies
         for (int j=i; j<endj; j++)
            this.bitstream.writeBits(frequencies[alphabet[j]], logMax);
      }

      return true;
   }

    
    // Reset frequency stats for each chunk of data in the block
    @Override
    public int encode(byte[] block, int blkptr, int len)
    {
       if ((block == null) || (blkptr + len > block.length) || (blkptr < 0) || (len < 0))
          return -1;
        
       if (len == 0)
          return 0;
      
       final int end = blkptr + len;
       final int sz = (this.chunkSize == 0) ? len : this.chunkSize;
       int startChunk = blkptr;

       while (startChunk < end)
       {         
           final int endChunk = (startChunk + sz < end) ? startChunk + sz : end;
           this.range = TOP_RANGE;
           this.low = 0;
           int lr = this.logRange;

           // Lower log range if the size of the data chunk is small
           while ((lr > 8) && (1<<lr > endChunk-startChunk))
              lr--;        
          
           if (this.rebuildStatistics(block, startChunk, endChunk, lr) < 0)
              return startChunk;
         
           this.shift = lr;
           
           for (int i=startChunk; i<endChunk; i++)
              this.encodeByte(block[i]);
          
           // Flush 'low'
           this.bitstream.writeBits(this.low, 60);
           startChunk = endChunk;
       }
       
       return len;
    }
    

    protected void encodeByte(byte b)
    {
        // Compute next low and range
        final int symbol = b & 0xFF;
        final long cumFreq = this.cumFreqs[symbol];
        final long freq = this.cumFreqs[symbol+1] - cumFreq;
        this.range >>>= this.shift;
        this.low += (cumFreq * this.range);
        this.range *= freq;
 
        // If the left-most digits are the same throughout the range, write bits to bitstream
        while (true)
        {
            if (((this.low ^ (this.low + this.range)) & RANGE_MASK) != 0)
            {
               if (this.range > BOTTOM_RANGE)
                  break;
               
               // Normalize
               this.range = -this.low & BOTTOM_RANGE;
            }

            this.bitstream.writeBits(this.low >> 40, 20);
            this.range <<= 20;
            this.low <<= 20;
        }
    }


   // Compute chunk frequencies, cumulated frequencies and encode chunk header
   private int rebuildStatistics(byte[] block, int start, int end, int lr)
   {
      for (int i=0; i<256; i++)
         this.freqs[i] = 0;

      for (int i=start; i<end; i++)
         this.freqs[block[i] & 0xFF]++;

      // Rebuild statistics
      return this.updateFrequencies(this.freqs, end-start, lr);      
   }
   
   
   @Override
    public OutputBitStream getBitStream()
    {
       return this.bitstream;
    }

   
    @Override
    public void dispose() 
    {
    }
}

File: cpp/src/entropy/RangeEncoder.cpp
/*
Copyright 2011-2017 Frederic Langlet
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
you may obtain a copy of the License at

                http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#include <sstream>
#include "RangeEncoder.hpp"
#include "../IllegalArgumentException.hpp"

using namespace kanzi;

// The chunk size indicates how many bytes are encoded (per block) before
// resetting the frequency stats. 0 means that frequencies calculated at the
// beginning of the block apply to the whole block.
// The default chunk size is 65536 bytes.
RangeEncoder::RangeEncoder(OutputBitStream& bitstream, int chunkSize, int logRange) THROW : _bitstream(bitstream)
{
    if ((chunkSize != 0) && (chunkSize < 1024))
        throw IllegalArgumentException("The chunk size must be at least 1024");

    if (chunkSize > 1 << 30)
        throw IllegalArgumentException("The chunk size must be at most 2^30");

    if ((logRange < 8) || (logRange > 16)) {
        stringstream ss;
        ss << "Invalid range parameter: " << logRange << " (must be in [8..16])";
        throw IllegalArgumentException(ss.str());
    }

    _logRange = logRange;
    _chunkSize = chunkSize;
    _shift = 0;
}

int RangeEncoder::updateFrequencies(uint frequencies[], int size, int lr)
{
    int alphabetSize = _eu.normalizeFrequencies(frequencies, _alphabet, 256, size, 1 << lr);

    if (alphabetSize > 0) {
        _cumFreqs[0] = 0;

        // Create histogram of frequencies scaled to 'range'
        for (int i = 0; i < 256; i++)
            _cumFreqs[i + 1] = _cumFreqs[i] + frequencies[i];
    }

    encodeHeader(alphabetSize, _alphabet, frequencies, lr);
    return alphabetSize;
}

bool RangeEncoder::encodeHeader(int alphabetSize, uint alphabet[], uint frequencies[], int lr)
{
    EntropyUtils::encodeAlphabet(_bitstream, alphabet, 256, alphabetSize);

    if (alphabetSize == 0)
        return true;

    _bitstream.writeBits(lr - 8, 3); // logRange
    int inc = (alphabetSize > 64) ? 16 : 8;
    int llr = 3;

    while (1 << llr <= lr)
        llr++;

    // Encode all frequencies (but the first one) by chunks of size 'inc'
    for (int i = 1; i < alphabetSize; i += inc) {
        uint max = 0;
        uint logMax = 1;
        int endj = (i + inc < alphabetSize) ? i + inc : alphabetSize;

        // Search for max frequency log size in next chunk
        for (int j = i; j < endj; j++) {
            if (frequencies[alphabet[j]] > max)
                max = frequencies[alphabet[j]];
        }

        while (uint(1 << logMax) <= max)
            logMax++;

        _bitstream.writeBits(logMax - 1, llr);

        // Write frequencies
        for (int j = i; j < endj; j++)
            _bitstream.writeBits(frequencies[alphabet[j]], logMax);
    }

    return true;
}

// Reset frequency stats for each chunk of data in the block
int RangeEncoder::encode(byte block[], uint blkptr, uint len)
{
    if (len == 0)
        return 0;

    const int end = blkptr + len;
    const int sz = (_chunkSize == 0) ? len : _chunkSize;
    int startChunk = blkptr;

    while (startChunk < end) {
        const int endChunk = (startChunk + sz < end) ? startChunk + sz : end;
        _range = TOP_RANGE;
        _low = 0;
        int lr = _logRange;

        // Lower log range if the size of the data chunk is small
        while ((lr > 8) && (1 << lr > endChunk - startChunk))
            lr--;

        if (rebuildStatistics(block, startChunk, endChunk, lr) < 0)
            return startChunk;

        _shift = lr;

        for (int i = startChunk; i < endChunk; i++)
            encodeByte(block[i]);

        // Flush 'low'
        _bitstream.writeBits(_low, 60);
        startChunk = endChunk;
    }

    return len;
}

inline void RangeEncoder::encodeByte(byte b)
{
    // Compute next low and range
    const int symbol = b & 0xFF;
    const uint64 cumFreq = _cumFreqs[symbol];
    const uint64 freq = _cumFreqs[symbol + 1] - cumFreq;
    _range >>= _shift;
    _low += (cumFreq * _range);
    _range *= freq;

    // If the left-most digits are the same throughout the range, write bits to bitstream
    while (true) {
        if (((_low ^ (_low + _range)) & RANGE_MASK) != 0) {
            if (_range > BOTTOM_RANGE)
                  break;

            // Normalize
            _range = ~(_low-1) & BOTTOM_RANGE;
        }

        _bitstream.writeBits(_low >> 32, 28);
        _range <<= 28;
        _low <<= 28;
    }
}

// Compute chunk frequencies, cumulated frequencies and encode chunk header
int RangeEncoder::rebuildStatistics(byte block[], int start, int end, int lr)
{
    memset(_freqs, 0, sizeof(_freqs));

    for (int i = start; i < end; i++)
        _freqs[block[i] & 0xFF]++;

    // Rebuild statistics
    return updateFrequencies(_freqs, end - start, lr);
}